packages/async-flow/src/log-store.js

```diff
@@ -0,0 +1,139 @@
+});
+
+/**
+ * A growable, replayable, sequence of `LogEntry`s.
+ *
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareLogStore = zone => {
+  /**
+   * @type {Ephemera<LogStore, {
+   *           index: number
+   *           replayDoneKit: import('@endo/promise-kit').PromiseKit<undefined>
+   *        }>}
+   */
+  const tmp = makeEphemera(log => {
+    const result = {
+      index: 0,
+      replayDoneKit: makePromiseKit(),
+    };
+    if (log.getLength() === 0) {
+      result.replayDoneKit.resolve(undefined);
+    }
+    return result;
+  });
+
+  return zone.exoClass(
+    'LogStore',
+    LogStoreI,
+    () => {
+      /**
+       * Really used to emulate a zone-storable vector, i.e., what in
+       * conventional JS you'd use a mutable array for, where you mutate
+       * only by `.push`
+       *
+       * @type {MapStore<number, LogEntry>}
+       */
+      const mapStore = zone.detached().mapStore('logMapStore', {
+        keyShape: M.number(),
+        valueShape: LogEntryShape,
+      });
+      return {
+        mapStore,
+      };
+    },
+    {
+      reset() {
+        const { self } = this;
+        tmp.resetFor(self);
+      },
+      dispose() {
+        const { state, self } = this;
+        tmp.resetFor(self);
+        state.mapStore.clear();
+      },
+      getIndex() {
+        const { self } = this;
+        return tmp.for(self).index;
+      },
+      getLength() {
+        const { state } = this;
+        return state.mapStore.getSize();
+      },
+      isReplaying() {
+        const { state, self } = this;
+        return tmp.for(self).index < state.mapStore.getSize();
+      },
+      peekEntry() {
+        const { state, self } = this;
+        self.isReplaying() ||
+          Fail`No longer replaying: ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        const result = state.mapStore.get(tmp.for(self).index);
+        return result;
+      },
+      nextEntry() {
+        const { self } = this;
+        const result = self.peekEntry();
+        tmp.for(self).index += 1;
+        if (!self.isReplaying()) {
+          tmp.for(self).replayDoneKit.resolve(undefined);
+        }
+        return result;
+      },
+      pushEntry(entry) {
+        const { state, self } = this;
+        !self.isReplaying() ||
+          Fail`still replaying: ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        tmp.for(self).index === state.mapStore.getSize() ||
+          Fail`internal: index confusion ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        state.mapStore.init(tmp.for(self).index, entry);
+        tmp.for(self).index += 1;
+        tmp.for(self).index === state.mapStore.getSize() ||
+          Fail`internal: index confusion ${q(tmp.for(self).index)} vs ${q(
+            state.mapStore.getSize(),
+          )}`;
+        return tmp.for(self).index;
```

Comment:
Are these repeated `tmp.for(self)` and `state.mapStore` calls really necessary?  ISTR "if you want a consistent answer, only ask once."
```suggestion
        const { state: { mapStore }, self } = this;
        const eph = tmp.for(self);
        !self.isReplaying() ||
          Fail`still replaying: ${q(eph.index)} vs ${q(mapStore.getSize())}`;
        eph.index === mapStore.getSize() ||
          Fail`internal: index confusion ${q(eph.index)} vs ${q(
            mapStore.getSize(),
          )}`;
        mapStore.init(eph.index, entry);
        eph.index += 1;
        eph.index === mapStore.getSize() ||
          Fail`internal: index confusion ${q(eph.index)} vs ${q(
            mapStore.getSize(),
          )}`;
        return eph.index;
```

Reply:
I used to write code like 
```js
const { state: { mapStore }, self } = this;
```

but `yarn format` turns that into
```js
        const {
          state: { mapStore },
          self,
        } = this;
```

so I took a suggestion from @turadg and only extract one level `{ state, self }` or `{ state, facets }` on that first line, and then work from there. But your other suggestions for reducing redundancy look good.

Reply:
Much better, thanks. Done.